src/pevm.rs

```diff
@@ -36,6 +32,13 @@ pub enum PevmError<C: PevmChain> {
     MissingTransactionData,
     /// Invalid input transaction.
     InvalidTransaction(C::TransactionParsingError),
+    /// Nonce too low or too high
+    NonceMismatch {
+        /// Nonce from tx (observed value)
```

Comment:
This is the `expected` value? 

A more helpful error would also include the transaction index. Maybe `tx_idx`, `expected_nonce`, and `observed_nonce`.

Reply:
The struct is "copied" from revm: https://github.com/risechain/revm/blob/main/crates/primitives/src/result.rs#L278-L285

```rust
    NonceTooHigh {
        tx: u64,
        state: u64,
    },
    NonceTooLow {
        tx: u64,
        state: u64,
    },
```

While adding `tx_idx` is good, using names `expected_nonce` and `observed_nonce` might be ambiguous. To me, `tx` is the "observed" value while `state` is the "expected" value. If this error happens, that's because the input is invalid (e.g. the block building service is feeding a nonce-too-low tx or nonce-to-high tx), not because the execution or state is wrong.

Reply:
Yeah, the problem is that either side can be wrong. A race condition in `pevm` or a corrupted state during canonical syncing would make `tx` the expected value instead. I got confused because `observed` is typically a calculated value. 

So yeah, let's do `tx_idx`, `tx_nonce`, and `calculated_nonce`/`executed_nonce`, and let the caller decide which is wrong.

Reply:
resolved